app/src/util_android.h

```diff
@@ -1127,6 +1127,12 @@ jint AttachCurrentThread(JavaVM* java_vm, JNIEnv** env);
 // App. If there is no instantiated App, returns nullptr.
 JNIEnv* GetJNIEnvFromApp();
 
+// Make sure the Java classes and methods are cached before triggering the
+// the callback. Can be slow if this is called BEFORE any Firebase App is
+// created.
+void CallAfterEnsureMethodsCached(JNIEnv* env, jobject activity,
```

Comment:
It's a little bit strange to me that this is implemented in app_android.cc and the header is util_android.h.

I think the implementation makes sense to be in app_android.cc as it calls CacheMethods and I guess there is not currently an app_android.h file, so I suppose this is fine as is.

Reply:
It is.

I'm not sure if it is worth adding a `app_android.h` just for this single function. And unfortunately the implementation has to live in `app_android.cpp` so that it can access to all the JNI caching and releasing functions.

Reply:
Why isn't the implementation in util_android, if it's not a method of App or related types like AppOptions?

Reply:
`CallAfterEnsureMethodsCached()` is an utility by nature but requires `CacheMethods()` and `ReleaseClasses()` which is only available in `app_android.cpp`. 

https://github.com/firebase/firebase-cpp-sdk/blob/main/app/src/app_android.cc

I can think of some alternative solutions
1. Add `app_android.h` and declare `CallAfterEnsureMethodsCached()` in `app_android.h`
1. Add `app_android.h` and declare `CacheMethods()` and `ReleaseClasses()` so that they can be linked to `util_android.cpp`. Then move `CallAfterEnsureMethodsCached()` to `util_android.cpp`.
1. Move declaration and impl of `CacheMethods()` and `ReleaseClasses()` to `util_android`.

Do you see a risk to implement a function declare in `util_android.h` but implemented in `app_android.cpp`?

Reply:
Added `app_android.h`